src/relay/backend/contrib/dnnl/codegen.cc

```diff
@@ -30,14 +30,108 @@
 #include <tvm/runtime/registry.h>
 
 #include <fstream>
+#include <numeric>
 #include <sstream>
 
+#include "../../utils.h"
 #include "../codegen_c/codegen_c.h"
 
 namespace tvm {
 namespace relay {
 namespace contrib {
 
+using namespace backend;
+
+const CallNode* GetRootConv2DCall(const CallNode* relu_call) {
+  CHECK(relu_call && IsOp(relu_call, "nn.relu"));
+  const auto relu_arg = relu_call->args[0];
+  const CallNode* add_call = relu_arg.as<CallNode>();
+  CHECK(add_call && IsOp(add_call, "add"));
+  const auto add_arg = add_call->args[0];
+  const CallNode* conv_call = add_arg.as<CallNode>();
+  CHECK(conv_call && IsOp(conv_call, "nn.conv2d"));
+  return conv_call;
+}
```

Comment:
Maybe this can be more generic, might get tedious to implement this kind of function if there are many different patterns. Could we just pass in a vector of op names like ["nn.relu", "add", "nn.conv2d"] along with the CallNode?

Reply:
yes exactly. I'm working on improving it right now, to support conv2d + relu pattern. I'm adding "depth" and "expected_op_names" arguments.

Reply:
fixed, please see the new change

Reply:
Sounds good to me. Backend I'm working on can use this, so it would be nice if this can move to utils.h where you moved IsOp and GetShape.

Reply:
Ok done. See the last commit.

Reply:
Thanks